File: src/lib/types/query.ts
export type CellValue =
  | { type: 'Null' }
  | { type: 'Bool'; value: boolean }
  | { type: 'Int'; value: number }
  | { type: 'Float'; value: number }
  | { type: 'Text'; value: string }
  | { type: 'Timestamp'; value: string }
  | { type: 'Binary'; value: number[] }
  | { type: 'Json'; value: string }
  | { type: 'LargeText'; value: { preview: string; full_length: number } }
  | { type: 'LargeJson'; value: { preview: string; full_length: number } }
  | { type: 'LargeBinary'; value: { preview_length: number; full_length: number } };

export interface ColumnDef {
  name: string;
  data_type: string;
}

export interface QueryResponse {
  columns: ColumnDef[];
  rows: CellValue[][];
  row_count: number;
  execution_time_ms: number;
  affected_rows: number | null;
  truncated?: boolean;
  max_rows_limit?: number;
}

export interface SortColumn {
  column: string;
  direction: 'ASC' | 'DESC';
}

export interface FilterCondition {
  column: string;
  operator: 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte' | 'contains' | 'starts_with' | 'is_null' | 'is_not_null';
  value: string;
}

export interface MultiStatementResult {
  results: QueryResponse[];
  error?: {
    index: number;
    message: string;
  };
}

export interface QueryHistoryEntry {
  id: string;
  connectionId: string;
  sql: string;
  executedAt: number;
  executionTimeMs: number;
  rowCount: number;
  error?: string;
}

export interface SavedQuery {
  id: string;
  name: string;
  connectionId: string;
  sql: string;
  createdAt: number;
  updatedAt: number;
}

File: src-tauri/src/models/query.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum CellValue {
    Null,
    Bool(bool),
    Int(i64),
    Float(f64),
    Text(String),
    Timestamp(String),
    Binary(Vec<u8>),
    Json(String),
    LargeText { preview: String, full_length: usize },
    LargeJson { preview: String, full_length: usize },
    LargeBinary { preview_length: usize, full_length: usize },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnDef {
    pub name: String,
    pub data_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryResponse {
    pub columns: Vec<ColumnDef>,
    pub rows: Vec<Vec<CellValue>>,
    pub row_count: usize,
    pub execution_time_ms: u64,
    pub affected_rows: Option<u64>,
    #[serde(default)]
    pub truncated: bool,
    #[serde(default)]
    pub max_rows_limit: Option<usize>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SortColumn {
    pub column: String,
    pub direction: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct FilterCondition {
    pub column: String,
    pub operator: String,
    pub value: String,
}
